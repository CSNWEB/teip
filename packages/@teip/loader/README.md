# Teip loader

Teip loader is an opinionated webpack loader that allows you to load and import `.graphql` files into your code.
