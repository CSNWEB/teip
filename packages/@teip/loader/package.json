{
  "name": "@teip/loader",
  "description": "A webpack loader for GraphQL files",
  "version": "0.1.0",
  "main": "bin/index.js",
  "repository": "https://github.com/hendrikniemann/teip/tree/master/packages/%40teip/loader",
  "author": "hendrikniemann",
  "license": "MIT",
  "private": false,
  "devDependencies": {
    "@babel/cli": "^7.0.0-beta.44",
    "@babel/core": "^7.0.0-beta.44",
    "@babel/preset-env": "^7.0.0-beta.44",
    "@babel/preset-flow": "^7.0.0-beta.44",
    "babel-jest": "^22.4.3",
    "flow-bin": "^0.69.0",
    "flow-typed": "^2.4.0",
    "jest": "^22.4.3"
  },
  "peerDependencies": {
    "graphql": "^0.13.2"
  },
  "babel": {
    "presets": [
      [
        "@babel/preset-env",
        {
          "targets": {
            "node": 8
          }
        }
      ],
      "@babel/preset-flow"
    ]
  },
  "scripts": {
    "build": "babel src/ --out-dir bin --ignore __test__",
    "prepublish": "npm run build"
  }
}
