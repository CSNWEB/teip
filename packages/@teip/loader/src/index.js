/* @flow */
import { parse } from 'graphql';
import { importRegex, transformImport, joinDocuments, findFragmentReferences } from './util';

export default function teipLoader(source: string) {
  const lines = source.split('\n');
  const lastImport = lines.findIndex(line => !line.match(importRegex));
  const imports = lines.slice(0, lastImport);
  const body = lines.slice(lastImport);
  const transformedImports = imports.map(transformImport);

  const { definitions } = parse(source);

  const transformedDefinition: string[] = definitions.map(definition => {
    let name = 'Unknown';
    const doc = {
      kind: 'Document',
      definitions: [definition],
    };
    const references = findFragmentReferences(doc).map(name => `${name}Fragment`);

    if (definition.kind === 'FragmentDefinition') {
      name = definition.name.value + 'Fragment';
    } else if (definition.kind === 'OperationDefinition') {
      if (!definition.name) {
        this.emitError(
          `One of your operations is missing a name. @teip/loader needs all operations to have names.`,
        );
        return '';
      }
      name =
        definition.name.value +
        definition.operation.charAt(0).toUpperCase() +
        definition.operation.slice(1);
    } else {
      this.emitError(
        `${
          this.resourcePath
        } contains definitions that are neither fragments nor operations. The @teip/loader can only handle fragments and operations!`,
      );
      return '';
    }

    if (references.length > 0) {
      return `var ${name} = joinDocuments(${references.join(', ')}, ${JSON.stringify(doc)});

export { ${name} };`;
    }
    return `var ${name} = ${JSON.stringify(doc)};

export { ${name} };`;
  });

  return `${transformedImports.join('\n')}

${joinDocuments.toString()}

${transformedDefinition.join('\n\n')}
`;
}
