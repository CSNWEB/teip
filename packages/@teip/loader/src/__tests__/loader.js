/* @flow */
import { parse } from 'graphql';
import loader from '../';

function createCompiler(callback: Function) {
  return {
    async: jest.fn(() => callback),
  };
}

describe('loader', () => {
  var compiler;
  var callback;

  it('should compile a simple file', () => {
    const source = `# import { Test } from './test.graphql'

query Hello {
  world {
    ...Test
  }
}
`;
    const compiler = createCompiler(() => {});
    const result = loader.call(compiler, source);
    expect(result).toMatchSnapshot();
  });

  it('should compile a file with multiple definitions', () => {
    const source = `
fragment Test on World {
  id
  denomination
}

query Hello {
  world {
    ...Test
  }
}
`;
    const compiler = createCompiler(() => {});
    const result = loader.call(compiler, source);
    expect(result).toMatchSnapshot();
  });

  it('should transpile files with mutations', () => {
    const source = `#import 'components/TransferShortList/definitions.graphql'

mutation doSomething {
  doSomething(what: "Something") {
    xyz
  }
}
`;

    const compiler = createCompiler(() => {});
    const result = loader.call(compiler, source);
    expect(result).toMatchSnapshot();
  });

  it('should transpile files with malformed imports', () => {
    const source = `#import 'components/TransferShortList/definitions.graphql'

query TransferShortList {
  transferChecklist {
    ...TransferShortListList
  }
  me {
    ...TransferShortListUser
  }
}
`;

    const compiler = createCompiler(() => {});
    const result = loader.call(compiler, source);
    expect(result).toMatchSnapshot();
  });
});
