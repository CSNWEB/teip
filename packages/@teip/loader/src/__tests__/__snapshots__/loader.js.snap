// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loader should compile a file with multiple definitions 1`] = `
"

function joinDocuments(...documents) {
  return {
    kind: 'Document',
    definitions: documents.map(function (doc) {
      return doc.definitions;
    }).reduce(function (acc, curr) {
      return acc.concat(curr);
    }, [])
  };
}

var TestFragment = {\\"kind\\":\\"Document\\",\\"definitions\\":[{\\"kind\\":\\"FragmentDefinition\\",\\"name\\":{\\"kind\\":\\"Name\\",\\"value\\":\\"Test\\",\\"loc\\":{\\"start\\":10,\\"end\\":14}},\\"typeCondition\\":{\\"kind\\":\\"NamedType\\",\\"name\\":{\\"kind\\":\\"Name\\",\\"value\\":\\"World\\",\\"loc\\":{\\"start\\":18,\\"end\\":23}},\\"loc\\":{\\"start\\":18,\\"end\\":23}},\\"directives\\":[],\\"selectionSet\\":{\\"kind\\":\\"SelectionSet\\",\\"selections\\":[{\\"kind\\":\\"Field\\",\\"name\\":{\\"kind\\":\\"Name\\",\\"value\\":\\"id\\",\\"loc\\":{\\"start\\":28,\\"end\\":30}},\\"arguments\\":[],\\"directives\\":[],\\"loc\\":{\\"start\\":28,\\"end\\":30}},{\\"kind\\":\\"Field\\",\\"name\\":{\\"kind\\":\\"Name\\",\\"value\\":\\"denomination\\",\\"loc\\":{\\"start\\":33,\\"end\\":45}},\\"arguments\\":[],\\"directives\\":[],\\"loc\\":{\\"start\\":33,\\"end\\":45}}],\\"loc\\":{\\"start\\":24,\\"end\\":47}},\\"loc\\":{\\"start\\":1,\\"end\\":47}}]};

export { TestFragment };

var HelloQuery = joinDocuments(TestFragment, {\\"kind\\":\\"Document\\",\\"definitions\\":[{\\"kind\\":\\"OperationDefinition\\",\\"operation\\":\\"query\\",\\"name\\":{\\"kind\\":\\"Name\\",\\"value\\":\\"Hello\\",\\"loc\\":{\\"start\\":55,\\"end\\":60}},\\"variableDefinitions\\":[],\\"directives\\":[],\\"selectionSet\\":{\\"kind\\":\\"SelectionSet\\",\\"selections\\":[{\\"kind\\":\\"Field\\",\\"name\\":{\\"kind\\":\\"Name\\",\\"value\\":\\"world\\",\\"loc\\":{\\"start\\":65,\\"end\\":70}},\\"arguments\\":[],\\"directives\\":[],\\"selectionSet\\":{\\"kind\\":\\"SelectionSet\\",\\"selections\\":[{\\"kind\\":\\"FragmentSpread\\",\\"name\\":{\\"kind\\":\\"Name\\",\\"value\\":\\"Test\\",\\"loc\\":{\\"start\\":80,\\"end\\":84}},\\"directives\\":[],\\"loc\\":{\\"start\\":77,\\"end\\":84}}],\\"loc\\":{\\"start\\":71,\\"end\\":88}},\\"loc\\":{\\"start\\":65,\\"end\\":88}}],\\"loc\\":{\\"start\\":61,\\"end\\":90}},\\"loc\\":{\\"start\\":49,\\"end\\":90}}]});

export { HelloQuery };
"
`;

exports[`loader should compile a simple file 1`] = `
"import { TestFragment } from './test.graphql';

function joinDocuments(...documents) {
  return {
    kind: 'Document',
    definitions: documents.map(function (doc) {
      return doc.definitions;
    }).reduce(function (acc, curr) {
      return acc.concat(curr);
    }, [])
  };
}

var HelloQuery = joinDocuments(TestFragment, {\\"kind\\":\\"Document\\",\\"definitions\\":[{\\"kind\\":\\"OperationDefinition\\",\\"operation\\":\\"query\\",\\"name\\":{\\"kind\\":\\"Name\\",\\"value\\":\\"Hello\\",\\"loc\\":{\\"start\\":47,\\"end\\":52}},\\"variableDefinitions\\":[],\\"directives\\":[],\\"selectionSet\\":{\\"kind\\":\\"SelectionSet\\",\\"selections\\":[{\\"kind\\":\\"Field\\",\\"name\\":{\\"kind\\":\\"Name\\",\\"value\\":\\"world\\",\\"loc\\":{\\"start\\":57,\\"end\\":62}},\\"arguments\\":[],\\"directives\\":[],\\"selectionSet\\":{\\"kind\\":\\"SelectionSet\\",\\"selections\\":[{\\"kind\\":\\"FragmentSpread\\",\\"name\\":{\\"kind\\":\\"Name\\",\\"value\\":\\"Test\\",\\"loc\\":{\\"start\\":72,\\"end\\":76}},\\"directives\\":[],\\"loc\\":{\\"start\\":69,\\"end\\":76}}],\\"loc\\":{\\"start\\":63,\\"end\\":80}},\\"loc\\":{\\"start\\":57,\\"end\\":80}}],\\"loc\\":{\\"start\\":53,\\"end\\":82}},\\"loc\\":{\\"start\\":41,\\"end\\":82}}]});

export { HelloQuery };
"
`;

exports[`loader should transpile files with malformed imports 1`] = `
"

function joinDocuments(...documents) {
  return {
    kind: 'Document',
    definitions: documents.map(function (doc) {
      return doc.definitions;
    }).reduce(function (acc, curr) {
      return acc.concat(curr);
    }, [])
  };
}

var TransferShortListQuery = joinDocuments(TransferShortListListFragment, TransferShortListUserFragment, {\\"kind\\":\\"Document\\",\\"definitions\\":[{\\"kind\\":\\"OperationDefinition\\",\\"operation\\":\\"query\\",\\"name\\":{\\"kind\\":\\"Name\\",\\"value\\":\\"TransferShortList\\",\\"loc\\":{\\"start\\":66,\\"end\\":83}},\\"variableDefinitions\\":[],\\"directives\\":[],\\"selectionSet\\":{\\"kind\\":\\"SelectionSet\\",\\"selections\\":[{\\"kind\\":\\"Field\\",\\"name\\":{\\"kind\\":\\"Name\\",\\"value\\":\\"transferChecklist\\",\\"loc\\":{\\"start\\":88,\\"end\\":105}},\\"arguments\\":[],\\"directives\\":[],\\"selectionSet\\":{\\"kind\\":\\"SelectionSet\\",\\"selections\\":[{\\"kind\\":\\"FragmentSpread\\",\\"name\\":{\\"kind\\":\\"Name\\",\\"value\\":\\"TransferShortListList\\",\\"loc\\":{\\"start\\":115,\\"end\\":136}},\\"directives\\":[],\\"loc\\":{\\"start\\":112,\\"end\\":136}}],\\"loc\\":{\\"start\\":106,\\"end\\":140}},\\"loc\\":{\\"start\\":88,\\"end\\":140}},{\\"kind\\":\\"Field\\",\\"name\\":{\\"kind\\":\\"Name\\",\\"value\\":\\"me\\",\\"loc\\":{\\"start\\":143,\\"end\\":145}},\\"arguments\\":[],\\"directives\\":[],\\"selectionSet\\":{\\"kind\\":\\"SelectionSet\\",\\"selections\\":[{\\"kind\\":\\"FragmentSpread\\",\\"name\\":{\\"kind\\":\\"Name\\",\\"value\\":\\"TransferShortListUser\\",\\"loc\\":{\\"start\\":155,\\"end\\":176}},\\"directives\\":[],\\"loc\\":{\\"start\\":152,\\"end\\":176}}],\\"loc\\":{\\"start\\":146,\\"end\\":180}},\\"loc\\":{\\"start\\":143,\\"end\\":180}}],\\"loc\\":{\\"start\\":84,\\"end\\":182}},\\"loc\\":{\\"start\\":60,\\"end\\":182}}]});

export { TransferShortListQuery };
"
`;

exports[`loader should transpile files with mutations 1`] = `
"

function joinDocuments(...documents) {
  return {
    kind: 'Document',
    definitions: documents.map(function (doc) {
      return doc.definitions;
    }).reduce(function (acc, curr) {
      return acc.concat(curr);
    }, [])
  };
}

var doSomethingMutation = {\\"kind\\":\\"Document\\",\\"definitions\\":[{\\"kind\\":\\"OperationDefinition\\",\\"operation\\":\\"mutation\\",\\"name\\":{\\"kind\\":\\"Name\\",\\"value\\":\\"doSomething\\",\\"loc\\":{\\"start\\":69,\\"end\\":80}},\\"variableDefinitions\\":[],\\"directives\\":[],\\"selectionSet\\":{\\"kind\\":\\"SelectionSet\\",\\"selections\\":[{\\"kind\\":\\"Field\\",\\"name\\":{\\"kind\\":\\"Name\\",\\"value\\":\\"doSomething\\",\\"loc\\":{\\"start\\":85,\\"end\\":96}},\\"arguments\\":[{\\"kind\\":\\"Argument\\",\\"name\\":{\\"kind\\":\\"Name\\",\\"value\\":\\"what\\",\\"loc\\":{\\"start\\":97,\\"end\\":101}},\\"value\\":{\\"kind\\":\\"StringValue\\",\\"value\\":\\"Something\\",\\"block\\":false,\\"loc\\":{\\"start\\":103,\\"end\\":114}},\\"loc\\":{\\"start\\":97,\\"end\\":114}}],\\"directives\\":[],\\"selectionSet\\":{\\"kind\\":\\"SelectionSet\\",\\"selections\\":[{\\"kind\\":\\"Field\\",\\"name\\":{\\"kind\\":\\"Name\\",\\"value\\":\\"xyz\\",\\"loc\\":{\\"start\\":122,\\"end\\":125}},\\"arguments\\":[],\\"directives\\":[],\\"loc\\":{\\"start\\":122,\\"end\\":125}}],\\"loc\\":{\\"start\\":116,\\"end\\":129}},\\"loc\\":{\\"start\\":85,\\"end\\":129}}],\\"loc\\":{\\"start\\":81,\\"end\\":131}},\\"loc\\":{\\"start\\":60,\\"end\\":131}}]};

export { doSomethingMutation };
"
`;
